password_enhancements.module

```diff
@@ -311,32 +354,23 @@ function password_enhancements_form_user_register_form_alter_validate(array &$fo
 
   $roles = $form_state->getValue('roles') ?: [];
 
-  // Try to load policy by it's role and priority.
-  try {
-    $entity_type_manager = \Drupal::entityTypeManager();
-    /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
-    $password_policy_storage = $entity_type_manager->getStorage('password_enhancements_policy');
-    $policy = $password_policy_storage->loadByRoleAndPriority($roles);
-
-    // Load constraints if there a policy exist for the given roles.
-    if ($policy !== NULL) {
-      /** @var \Drupal\password_enhancements\Entity\Storage\ConstraintEntityStorageInterface $constraint_storage */
-      $constraint_storage = $entity_type_manager->getStorage('password_enhancements_constraint');
-      $constraints = $constraint_storage->loadByRoles($roles);
-    }
-  }
-  catch (\Exception $e) {
-    \Drupal::service('logger.password_enhancements')->logException('Failed to load policy and constraints.', $e);
-    $form_state->setError($form['account']['pass'], t('Password validation error: failed to load policy and constraints.<br>Changing the password is not possible at the moment, please try again later.'));
+  // Try to load policy by it's role and priority and bail out early if no
+  // applicable policy is found.
+  /** @var \Drupal\password_enhancements\PasswordPolicyInterface $policy */
+  $policy = PasswordPolicy::loadByRoles($roles);
+  if ($policy === NULL) {
     return;
   }
 
+  // Load the constraints for the given roles.
+  $constraints = _password_enhancements_load_constraints_by_roles($roles);
```

Comment:
And this calls `PasswordPolicy::loadByRoles($roles);` again. Even if that method does some kind of caching, would not be easier to bail out early in that function (hopefully method in the future) and here as well if `empty($constraints)`?

Reply:
I'm not sure I'm getting what you mean here: that's exactly what happens. What change would you expect?

Reply:
It was 2 weeks ago, but I guess what I wanted to suggest is instead of calling `PasswordPolicy::loadByRoles($roles)` above that calls `PasswordPolicy::loadMultipleByRoles()` (that performs expensive actions: filtering and sorting without static caching) and then calling `PasswordPolicy::loadMultipleByRoles()` **again** in `_password_enhancements_load_constraints_by_roles()` if  `$policy !== NULL` was above should be avoided.

I would also avoid static caching in service because that could also cause some headaches. So maybe you should call `PasswordPolicy::loadMultipleByRoles()` instead of `PasswordPolicy::loadByRoles($roles)` and have a ~function~ method that could load constraints from what `PasswordPolicy::loadMultipleByRoles()` returns.